csharp/src/Ice/ConnectionI.cs

```diff
@@ -2643,49 +2632,93 @@ private int parseMessage(ref MessageInfo info)
         {
             while (requestCount > 0)
             {
-                //
-                // Prepare the invocation.
-                //
-                bool response = !_endpoint.datagram() && requestId != 0;
-                Debug.Assert(!response || requestCount == 1);
-
-                inc = getIncoming(adapter, response, compress, requestId);
-
-                //
-                // Dispatch the invocation.
-                //
-                inc.dispatch(servantManager, stream);
+                // adapter can be null here, however the adapter set in current can't be null, and we never pass
+                // a null current.adapter to the application code. Once this file enables nullable, adapter should be
+                // adapter! below.
+                var request = new IncomingRequest(requestId, this, adapter, stream);
 
+                if (dispatcher is not null)
+                {
+                    // We don't and can't await the dispatchAsync: with batch requests, we want all the dispatches to
+                    // execute in the current Ice thread pool thread. If we awaited the dispatchAsync, we could
+                    // switch to a .NET thread pool thread.
+                    _ = dispatchAsync(request);
+                }
+                else
+                {
+                    // Received request on a connection without an object adapter.
+                    bool isTwoWay = !_endpoint.datagram() && requestId != 0;
+                    if (isTwoWay)
+                    {
+                        sendResponse(
+                            requestId,
+                            request.current.createOutgoingResponse(new ObjectNotExistException()).outputStream,
+                            0,
+                            amd: false);
+                    }
+                    else
+                    {
+                        sendNoResponse();
+                    }
+                }
                 --requestCount;
-
-                reclaimIncoming(inc);
-                inc = null;
             }
 
             stream.clear();
         }
-        catch (LocalException ex)
+        catch (LocalException ex) // TODO: catch all exceptions
         {
-            this.dispatchException(requestId, ex, requestCount, false);
+            this.dispatchException(requestId, ex, requestCount, amd: false);
         }
-        finally
+
+        async Task dispatchAsync(IncomingRequest request)
         {
-            if (inc != null)
+            bool isTwoWay = !_endpoint.datagram() && requestId != 0;
+            bool amd = false;
+
+            try
             {
-                reclaimIncoming(inc);
+                OutgoingResponse response;
+
+                try
+                {
+                    ValueTask<OutgoingResponse> valueTask = dispatcher.dispatchAsync(request);
+                    amd = !valueTask.IsCompleted;
+                    response = await valueTask.ConfigureAwait(false);
```

Comment:
Wondering if the sync dispatch can be optimized with:

```
ValueTask<OutgoingResponse> valueTask = dispatcher.dispatchAsync(request);
if (valueTask.IsCompletedSuccessfully)
{
    response = valueTask.Result;
}
else
{
    amd = true;
    response = await valueTask.ConfigureAwait(false);
}
```

The actual code is simpler, so is probably fine to keep it as is.